src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -452,28 +469,45 @@ - (void)invokeCommandWithEndpointID:(NSNumber *)endpointID
-              expectedValueInterval:(NSNumber *)expectedValueInterval
+                     expectedValues:(NSArray<NSDictionary<NSString *, id> *> * _Nullable)expectedValues
+              expectedValueInterval:(NSNumber * _Nullable)expectedValueInterval
                  timedInvokeTimeout:(NSNumber * _Nullable)timeout
                         clientQueue:(dispatch_queue_t)clientQueue
                          completion:(MTRDeviceResponseHandler)completion
 {
-    // Perform this operation
-    MTRBaseDevice * baseDevice = [[MTRBaseDevice alloc] initWithNodeID:self.nodeID controller:self.deviceController];
-    [baseDevice
-        invokeCommandWithEndpointId:endpointID
-                          clusterId:clusterID
-                          commandId:commandID
-                      commandFields:commandFields
-                 timedInvokeTimeout:timeout
-                        clientQueue:self.queue
-                         completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
-                             dispatch_async(clientQueue, ^{
-                                 completion(values, error);
-                             });
-                         }];
-
-    [self setExpectedValues:expectedValues expectedValueInterval:expectedValueInterval];
+    if (timeout) {
+        timeout = MTRClampedNumber(timeout, @(1), @(UINT16_MAX));
+    }
+    if (expectedValueInterval) {
+        if ([expectedValueInterval compare:@(0)] == NSOrderedAscending) {
+            expectedValues = nil;
+        } else {
+            expectedValueInterval = MTRClampedNumber(expectedValueInterval, @(1), @(UINT32_MAX));
+        }
+    }
+    MTRAsyncCallbackQueueWorkItem * workItem = [[MTRAsyncCallbackQueueWorkItem alloc] initWithQueue:_queue];
+    MTRAsyncCallbackReadyHandler readyHandler = ^(MTRDevice * device, NSUInteger retryCount) {
+        MTRBaseDevice * baseDevice = [self newBaseDevice];
+        [baseDevice
+            invokeCommandWithEndpointId:endpointID
+                              clusterId:clusterID
+                              commandId:commandID
+                          commandFields:commandFields
+                     timedInvokeTimeout:timeout
+                            clientQueue:self.queue
+                             completion:^(NSArray<NSDictionary<NSString *, id> *> * _Nullable values, NSError * _Nullable error) {
+                                 dispatch_async(clientQueue, ^{
+                                     completion(values, error);
+                                 });
+                                 [workItem endWork];
+                             }];
+    };
+    workItem.readyHandler = readyHandler;
+    [_asyncCallbackWorkQueue enqueueWorkItem:workItem];
+
+    if (expectedValues) {
+        [self setExpectedValues:expectedValues expectedValueInterval:expectedValueInterval];
```

Comment:
So if we get here with `expectedValues` not nil but `expectedValueInterval` nil, what happens?  `setExpectedValues` seems to expect a non-nil interval, at first glance....

Reply:
That would be an API usage error. I will add a change to nil expectedValues and log.

I'm also wondering if the API should respond more drastically like throw an exception? In this case I feel it may be fine because nothing operational would be broken - just that the attribute report will come a little later, in the usual case.

Reply:
> That would be an API usage error.

OK.  We should document that constraint in the API docs (they don't have it right now) and document what the behavior is if it's violated...

> I'm also wondering if the API should respond more drastically like throw an exception?

I would be fine with that.  We just need to make sure we don't crash in our own code here.

Reply:
I'm keeping the logic to ignore the values if either argument is nil or <= 0, since the behavior will be "works eventually".